### DISCUSSION TITLE
Leap

### DISCUSSION THREAD

**Discussion by meepleek** [71/dance#278] [Q&A]
The leap impl is bleeding edge, so I'm mostly curious about the planned features for this awesome feature.
Is bidirectional leap planned? Could leap extend be a thing or is that just a silly idea?

Are unusable suggestions supposed to stay (e.g. 'pe' when leaping to 'pr' in the screenshot? Also I'm getting very small label groups so that even with just a few matches there are 2 groups - how are the labels picked?
![image](https://user-images.githubusercontent.com/17273703/196009985-7f531f07-0fe1-4be6-aa02-6e4e3e8e19ba.png)

**Comment by 71** [comment]
Thanks for trying it out!

I'm actually open to suggestions for the `leap` API, given that the original one was designed for Nvim, but Kakoune is quite different. Extend `leap` could make sense and can probably be added quite easily (on that note, should default `leap` "jump" to the destination, or "select up" to the destination?). In any case, I'm not sure whether we can make `leap` work with multiple selections.

`pa` staying in the screenshot above is a bug, I'll look into it.

For now the available labels are hardcoded to avoid conflicting with other keybindings after jumping; if you have suggestions please let me know!

https://github.com/71/dance/blob/73d32adb80497758f48a643266ba3488d0c08561/src/commands/seek.ts#L512

**Reply by meepleek** [reply]
I've just started learning modal editing and I'm on a custom layout (a colemak-ish 40%), so take this with a grain of salt.

I wouldn't mind having labels for all matches including the closest one which might allow for a leap mode to temporarily take over to prevent conflicts. In fact the nvim leap does this for the bidirectional leap (every match has labels). 
My reasoning for bidirectional is that it might further lower the cognitive load by not caring where the target is relative to the cursor.

I meant select when I wrote extend, sorry.

As for the labels, it would be super nice if those were configurable and the order would be taken into the consideration to account for different layouts. (E.g. in my case I would start with 'arstneio' as that's the colemak homerow, but would omit 'qx' as I have those on a different layer due to their low frequency).

**Reply by meepleek** [reply]
I just noticed the fn already takes the labels as a param, which is nice.

I also forgot to ask: is leaping across splits is planned or even possible for that matter?

**Comment by tshort** [comment]
Cool feature!!
